Remotion/Validation/UnitTests/DefaultValidatorExtensionsTest.cs

```diff
@@ -1,14 +1,416 @@
+      var lengthValidator = (LengthValidator)createdValidator;
+      Assert.That(lengthValidator.Min, Is.EqualTo(17));
+      Assert.That(lengthValidator.Max, Is.EqualTo(42));
+    }
+
+    [Test]
+    public void MinLength_ForString_ReturnsLengthValidator ()
+    {
+      var validationMessage = new InvariantValidationMessage("Fake message");
+      var initParameters = new PropertyValidationRuleInitializationParameters(validationMessage);
+      IPropertyValidator createdValidator = null;
+      var addingPropertyValidationRuleBuilder = new Mock<IAddingPropertyValidationRuleBuilder<object, string>>();
+      addingPropertyValidationRuleBuilder.Setup(_ => _.SetValidator(It.IsAny<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>()))
+          .Callback<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>(func => createdValidator = func.Invoke(initParameters));
+
+      addingPropertyValidationRuleBuilder.Object.MinLength(17);
+
+      addingPropertyValidationRuleBuilder.Verify();
+      Assert.That(createdValidator, Is.InstanceOf<LengthValidator>());
+      var lengthValidator = (LengthValidator)createdValidator;
+      Assert.That(lengthValidator.Min, Is.EqualTo(17));
+      Assert.That(lengthValidator.Max, Is.Null);
+    }
+
+    [Test]
+    public void MaxLength_ForString_ReturnsLengthValidator ()
+    {
+      var validationMessage = new InvariantValidationMessage("Fake message");
+      var initParameters = new PropertyValidationRuleInitializationParameters(validationMessage);
+      IPropertyValidator createdValidator = null;
+      var addingPropertyValidationRuleBuilder = new Mock<IAddingPropertyValidationRuleBuilder<object, string>>();
+      addingPropertyValidationRuleBuilder.Setup(_ => _.SetValidator(It.IsAny<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>()))
+          .Callback<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>(func => createdValidator = func.Invoke(initParameters));
+
+      addingPropertyValidationRuleBuilder.Object.MaxLength(42);
+
+      addingPropertyValidationRuleBuilder.Verify();
+      Assert.That(createdValidator, Is.InstanceOf<LengthValidator>());
+      var lengthValidator = (LengthValidator)createdValidator;
+      Assert.That(lengthValidator.Min, Is.EqualTo(0));
+      Assert.That(lengthValidator.Max, Is.EqualTo(42));
+    }
+
+    [Test]
+    public void ExactLength_ForString_ReturnsLengthValidator ()
+    {
+      var validationMessage = new InvariantValidationMessage("Fake message");
+      var initParameters = new PropertyValidationRuleInitializationParameters(validationMessage);
+      IPropertyValidator createdValidator = null;
+      var addingPropertyValidationRuleBuilder = new Mock<IAddingPropertyValidationRuleBuilder<object, string>>();
+      addingPropertyValidationRuleBuilder.Setup(_ => _.SetValidator(It.IsAny<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>()))
```

Comment:
Naming: please suffix those test doubles to keep with the existing naming conventions.
...Stub for stub
...Mock for mock

Here, it's all stubs since we don't do any expections.

Reply:
`addingPropertyValidationRuleBuilder` uses a `Setup` (which creates an expectation) and the call to `SetValidator` is also verified. Is it still a Stub? Should I remove the call to `Verify()`?

Reply:
`Setup` is used for both stub and mock. Only by adding `Verifiable` at the end, it becomes mock semantics.

Since making a mock is only needed when we want to assert that void-calls are done correctly, we can stick with the existing stub semantics. The storing of the created validator during the callback and later asserting that value takes care of the verifications we need to do.

Reply:
I see. I changed the names to "...Stub" and removed the calls to `Verify()`, except for the `predicateMock`, which I made `Verifiable` in order to check that the predicate I passed as a parameter is actually the one used in the `createdValidator`.